### PR TITLE
Fix error reporting context for missing generic type arguments

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/7077

The logic of the fix is quite straightforward: emit the error immediately when we fix an instance type, and not in a traversal after the main pass. Note that I fix this only for the new analyser, because the old one will be anyway deprecated soon.

This causes a bit of code churn because we need to pass the flags to several functions, but the logic should not change much. I didn't add many new tests, since we have a bunch of existing tests. It would probably make sense to play with this, suggestions for more tests are very welcome.

### CODE REVIEW

**Comment by msullivan** [comment]
I wasn't able to fully test this out internally because I ran into some other issues but one thing I noticed is this makes aliases like `Foo = List` illegal. (Even though if you `# type: ignore` the error they do work correctly)

File: mypy/newsemanal/semanal.py [hunk 1]
```diff
@@ -2369,7 +2369,9 @@ def check_and_set_up_type_alias(self, s: AssignmentStmt) -> bool:
         # so we need to replace it with non-explicit Anys.
         res = make_any_non_explicit(res)
         no_args = isinstance(res, Instance) and not res.args
-        fix_instance_types(res, self.fail)
+        fix_instance_types(res, self.fail,
```
Comment (msullivan) [inline mypy/newsemanal/semanal.py]: I think this needs to be done conditionally only when `no_args`, maybe?

Comment (ilevkivskyi) [reply mypy/newsemanal/semanal.py]: Yes, aliases like `A = List` should be allowed even with the flag, will fix it now.

**Comment by ilevkivskyi** [comment]
@msullivan I fixed the errors you reported, could you please try again?

**Pull request merged**